src/GridFS/GridFsUpload.php

```diff
@@ -0,0 +1,169 @@
+     * @param string                 $filename         Filename to insert
+     * @param array                  $options          File options
+     * @throws InvalidArgumentException
+     */
+    public function __construct(
+        Bucket $bucket,
+        $filename,
+        array $options=[]
+        )
+    {
+        $this->bufferLength = 0;
+        $this->ctx = hash_init('md5');
+
+        $uploadDate = time();
+        $objectId = new \MongoDB\BSON\ObjectId();
+        $main_file = [
+            "chunkSize" => $bucket->getChunkSizeBytes(),
+            "filename" => $filename,
+            "uploadDate" => $uploadDate,
+            "_id" => $objectId
+        ];
+
+        $fileOptions = [];
+        if (isset($options['contentType'])) {
+            if (is_string($options['contentType'])) {
+                $fileOptions['contentType'] = $options['contentType'];
+            } else {
+                throw new InvalidArgumentTypeException('"contentType" option', $options['contentType'], 'string');
+            }
+        }
+        if (isset($options['aliases'])) {
+            if (\MongoDB\is_string_array($options['aliases'])) {
+                $fileOptions['aliases'] = $options['aliases'];
+            } else {
+                throw new InvalidArgumentTypeException('"aliases" option', $options['aliases'], 'array of strings');
+            }
+        }
+
+        if (isset($options['metadata'])) {
+            if (is_array($options['metadata']) || is_object($options['metadata'])) {
+                $fileOptions['metadata'] = $options['metadata'];
+            } else {
+                throw new InvalidArgumentTypeException('"metadata" option', $options['metadata'], 'object or array');
+            }
+        }
+        $this->file = array_merge($main_file, $fileOptions);
+        parent::__construct($bucket);
+    }
+    /**
+    * Reads data from a stream into GridFS
+    *
+    * @param Stream   $source   Source Stream
+    * @return ObjectId
+    */
+    public function uploadFromStream($source)
+    {
+        $this->bucket->ensureIndexes();
+
+        if (!is_resource($source) || get_resource_type($source) != "stream") {
+            throw new UnexpectedTypeException('stream', $source);
+        } else{
+            $streamMetadata = stream_get_meta_data($source);
+        } if (!is_readable($streamMetadata['uri'])) {
+     //       throw new InvalidArgumentException("stream not readable");
+            //issue being that php's is_readable reports native streams as not readable like php://temp
+        }
+        while ($data = fread($source, $this->bucket->getChunkSizeBytes())) {
+            $this->insertChunk($data);
+        }
+        return $this->fileCollectionInsert();
+    }
+    /**
+    * Insert a chunks into GridFS from a string
+    *
+    * @param string  $toWrite   String to upload
+    * @return int
+    */
+    public function insertChunks($toWrite)
+    {
+        $this->bucket->ensureIndexes();
```

Comment:
So this is actually part of the reason for my refactor. As soon as fopen() is called within openUploadStream, the bucket is no longer actually really able to see what is going on so the stream needs to sort of go around it. I could change the bucket class to have methods that wrap writes to its underlying db's where it would be called.